### PR TITLE
Add a multipart copy abstraction

### PR DESCRIPTION
**Pull request by jeskew**
This PR allows for copying large objects. It uses the multipart abstraction set up for Glacier and S3 uploads, but I had to refactor `Aws\Multipart\AbstractUploader` to isolate interaction with streams, as no streams are involved in a multipart copy. The [first commit](https://github.com/jeskew/aws-sdk-php/commit/67fb8ee1e35161f4e8d5c8893c78d0f85216f2d7) is just refactoring all the non-stream stuff into `Aws\Multipart\AbstractUploadManager`, from which `Aws\Multipart\AbstractUploader` descends.

The [second commit](https://github.com/jeskew/aws-sdk-php/commit/55a46135a116f6695a884190516e21636bbd5260) adds `Aws\S3\MultipartCopy`, which is a descendent of  `Aws\Multipart\AbstractUploadManager`. It also adds a `copy` method to `Aws\S3\S3Client`, which calls `headObject` on a source and then performs either a `copyObject` operation or a multipart copy, depending on the source's size. `Aws\S3\StreamWrapper::rename` now calls `copy` instead of `copyObject` to support objects larger than 5GB.

This would address #759.

/cc @mtdowling

### CODE REVIEW

**Comment by lonormaly** [comment]
This is more than great! When will the PR be merged? Can't wait to test it :)

**Comment by jeskew** [comment]
@lonormaly It still needs to be reviewed. You can check out the feature branch directly if you'd like to do some exploratory testing.

**Comment by jeskew** [comment]
@mtdowling The latest commit adds support for copying versions of objects, as that portion of the key [cannot be URL encoded](https://github.com/aws/aws-sdk-js/issues/727).

File: src/Multipart/AbstractUploadManager.php [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+use GuzzleHttp\Promise;
+use GuzzleHttp\Promise\PromiseInterface;
+use GuzzleHttp\Psr7;
+use InvalidArgumentException as IAE;
+use Psr\Http\Message\RequestInterface;
+
+/**
+ * Encapsulates the execution of a multipart upload to S3 or Glacier.
+ */
+abstract class AbstractUploadManager implements Promise\PromisorInterface
```
Comment (mtdowling) [inline src/Multipart/AbstractUploadManager.php]: Mark as `@internal`?

**Comment by mtdowling** [comment]
This looks good. The only thing that concerns me is making a bunch of stuff protected, but if we mark that class as `@internal`, I think this can be :ship:'d

**Comment by lonormaly** [comment]
Please! This one is really important for us :+1:

**Pull request merged**